types/context.go

```diff
@@ -16,6 +16,20 @@ import (
 	"cosmossdk.io/core/header"
 )
 
+// ExecMode defines the execution mode which can be set on a Context.
+type ExecMode uint8
+
+// All possible execution modes.
+const (
+	ExecModeCheck ExecMode = iota
+	ExecModeReCheck
+	ExecModeSimulate
+	ExecModePrepareProposal
+	ExecModeProcessProposal
+	ExecModeVoteExtension
+	ExecModeFinalize
+)
```

Comment:
should we import baseapp here instead of duplicating?

Reply:
We can't due to cyclic dependencies (`baseapp` <-> `types`). I'm tempted to actually keep them here and have BaseApp depend on them. WDYT?

Reply:
this is hard, maybe we leave it for now and do the refactor as part of runtime/baseapp refactor.